great_tables/_gt_data.py

```diff
@@ -514,58 +527,108 @@ class RowInfo:
-                group_id = to_list(data[groupname_col])
-            else:
-                group_id = [None] * n_rows(data)
+        if rowname_col is not None:
+            row_names = to_list(data[rowname_col])
+        else:
+            row_names = [None] * n_rows(data)
 
-            if rowname_col is not None:
-                row_names = to_list(data[rowname_col])
-            else:
-                row_names = [None] * n_rows(data)
+        # Obtain the column names from the data and initialize the
+        # `_stub` from that
+        row_info = [RowInfo(*i) for i in zip(row_indices, group_id, row_names)]
 
-            # Obtain the column names from the data and initialize the
-            # `_stub` from that
-            self._d = [RowInfo(*i) for i in zip(row_indices, group_id, row_names)]
+        # create groups, and ensure they're ordered by first observed
+        group_names = list({row.group_id: True for row in row_info if row.group_id is not None})
+        group_rows = GroupRows(data, group_key=groupname_col).reorder(group_names)
+
+        return cls(row_info, group_rows)
+
+    def _set_cols(
+        self, data: TblData, boxhead: Boxhead, rowname_col: str | None, groupname_col: str | None
+    ) -> Tuple[Stub, Boxhead]:
+        """Return a new Stub and Boxhead, with updated rowname and groupname columns.
+
+        Note that None unsets a column.
+        """
+
+        new_boxhead = boxhead.set_stub_cols(rowname_col, groupname_col)
+        new_stub = self.from_data(data, rowname_col, groupname_col)
+
+        return new_stub, new_boxhead
+
+    @property
+    def group_ids(self) -> RowGroups:
+        return [group.group_id for group in self.group_rows]
+
+    def reorder_rows(self, indices) -> Self:
+        new_rows = [self.rows[ii] for ii in indices]
 
-    def _to_row_groups(self) -> RowGroups:
-        # get unique group_ids, using dict as an ordered set
-        group_ids = list({row.group_id: True for row in self if row.group_id is not None})
+        return self.__class__(new_rows, self.group_rows)
 
-        return group_ids
+    def order_groups(self, group_order: RowGroups):
+        # TODO: validate
+        return self.__class__(self.rows, self.group_rows.reorder(group_order))
+
+    def group_indices_map(self) -> list[tuple[int, str | None]]:
+        return self.group_rows.indices_map(len(self.rows))
+
+    def __iter__(self):
+        return iter(self.rows)
+
+    def __len__(self):
+        return len(self.rows)
+
+    def __getitem__(self, ii: int):
+        return self.rows[ii]
 
     def _get_stub_components(self) -> list[str]:
         stub_components: list[str] = []
 
-        if any(entry.group_id is not None for entry in self):
+        if any(entry.group_id is not None for entry in self.rows):
             stub_components.append("group_id")
 
-        if any(entry.rowname is not None for entry in self):
+        if any(entry.rowname is not None for entry in self.rows):
```

Comment:
@machow , I'm just browsing through the code. It looks great, but I suspect you could iterate over `self` directly without changing these lines, since `__iter__` is already implemented.

Reply:
Ah, nice catch! Since Stub now has two pieces inside it (rows and group_rows), I wasn't sure whether it should allow iterating over it, or require grabbing rows directly. I think the `__iter__` was to avoid breaking code while refactoring. WDYT, is it worth keeping `__iter__`?

Reply:
I see. Personally, I would prefer keeping `__iter__`.